doc/common/Cadl-Project-Scripts.md

```diff
@@ -0,0 +1,127 @@
+
+Example
+
+```powershell
+function Get-dotnet-EmitterPackageJsonPath() {
+  return "D:\SomeOtherLocation\some-other-emitter-package.json"
+}
+```
+
+#### Get-${Language}-EmitterAdditionalOptions (Optional)
+
+This function allows you to append additional `--option` arguments that will be passed into cadl compile.  One example of this is the `emitter-output-dir`.  For dotnet we want the location of the generated files to be `{projectDir}/src` however in other languages they will have other conventions.  This method will take in a fully qualified path to the project directory so you can construct your relative path to that as the output.
+
+Example
+
+```powershell
+function Get-dotnet-EmitterAdditionalOptions([string]$projectDirectory) {
+  return "--option @azure-tools/cadl-csharp.emitter-output-dir=$projectDirectory/src"
+}
+```
+
+## Cadl-Project-Sync.ps1
+
+This is the first script that should be called and can be found at `./eng/common/scripts/Cadl-Project-Sync.ps1`.  It takes in one parameter which is the root directory of the project which is typically one layer lower than the service directory.  As an example for dotnet this is `./sdk/openai/Azure.AI.OpenAI` where `openai` is the service directory and `Azure.AI.OpenAI` is the project directory.
+
+This script will create a sparse check out at the root of your current repository named after the project directory, automatically filter to only the files in the directory defined in cadl-location.yaml, and sync to the sha defined in cadl-location.yaml.
+
+If you have your language repo at `D:\git\azure-sdk-for-net` there will be a new directory `D:\git\sparse-checkout\Azure.AI.OpenAI` where the sparse checkout will live.
+
+This is then copied over to your project directory so that you can make temporary changes if needed.  The location will be `./{projectDir}/TempCadlFiles`.  This temporary directory will be cleaned up at the end of the generate script if set in the cadl-location.yaml.
+
+## Cadl-Project-Generate.ps1
+
+This is the second script that should be called and can be found at `./eng/common/scripts/Cadl-Project-Generate.ps1`.  It takes the exact same parameter as the sync script.
+
+The first thing this does is clean up the npm install that might exist in `./{projectDir}/TempCadlFiles`, followed by replacing the package.json with the language static one.
+
+Once this is done it will run `npm install` followed by `cadl compile` which is the standard way to generate a cadl project.
+
+The exact command that gets run is output stdout to enable debugging if needed.
+
+We currently don't do anything to the cadl-project.yaml that gets pulled in from the spec repo to limit to just your language emitter instead we use the filter option on the command line `--emit $emitterName`.  This allows you to isolate the generation to only things owned by your language so you can safely add generation dependencies in CI without needing to worry about noisy neighbors.
```

Comment:
When running `cadl compile` directly, I can point to `client.cadl` instead of `main.cadl`. I'd expect that this would be a fairly common scenario as we'll have some customizations for the client libraries. Will this script always look for `main.cadl` from the sparsely cloned directory? Is there a way to tell the script to use client.cadl instead?

Reply:
good point, today yes we look for main.cadl.

I see this as a static setting for a given project, do you see this more as a cadl-location.yaml setting or a cadl-project.yaml setting?  I lean towards cadl-location.yaml only because this changes what to pass into cadl, and cadl is what reads the cadl-project.yaml so it seems chicken / egg.

Reply:
`cadl-location.yaml` is a good place to configure this.

Reply:
I just saw that is what we were doing so we can probably resolve this https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/scripts/Cadl-Project-Generate.ps1#L64